TypeScript/app/gilded-rose.ts

```diff
@@ -10,6 +12,127 @@ export class Item {
   }
 }
 
+interface ItemUpdateStrategy {
+  /**
+   * Determines whether the specified item qualifies for this strategy.
+   * @param item The item to check.
+   * @returns `true` if the item qualifies, `false` otherwise.
+   */
+  isForItem(item: Item): boolean;
+
+  /**
+   * Updates the item quality and sellIn values according to the strategy rules.
+   * @param item The item to update.
+   */
+  updateItem(item: Item): void;
+}
+
+/**
+ * The list of supported item update strategies.
+ *
+ * IMPORTANT: The order of items in this array is important. Several
+ * update strategies can return `true` from `isForItem()` for
+ * the same item. The first update strategy to return `true` will
+ * be used.
+ */
+export const ITEM_UPDATE_STRATEGIES: ItemUpdateStrategy[] = [
```

Comment:
Each strategy can be implement as a proper `class` and even benefit from inheritance (e.g. almost all of them decrease `sellIn`, which can be implemented in the base class). I decided to keep them separate for the sake of simplicity and overall tendency to not use full-blown classes in JS/TS.